source/agora/crypto/Bech32.d

```diff
@@ -0,0 +1,486 @@
+
+char[] encode_bech32 (string hrp, ubyte[] values)
+{
+    assert(values !is null);
+
+    char[] encoded;
+    ubyte[] conv;
+    convertBits(conv, values, 8, 5, true);
+
+    // First ensure that the HRP is all lowercase. BIP-173 requires an encoder
+    // to return a lowercase Bech32 string, but if given an uppercase HRP, the
+    // result will always be invalid.
+    foreach (const ref c; hrp)
+        assert(c < 'A' || c > 'Z');
+
+    ubyte[] checksum = create_checksum(hrp, conv);
+    conv ~= checksum;
+
+    uint index = 0;
+    encoded ~= hrp ~ "1";
+    foreach (const c ; conv)
+        encoded ~= CHARSET[c];
+
+    return encoded;
+}
+
+/*******************************************************************************
+
+    Decode a Bech32 string.
+
+    Params:
+        str = encoded string
+
+    Returns:
+        decoded ubyte array
+
+*******************************************************************************/
+
+ubyte[] decode_bech32 (const char[] str)
+{
+    import std.string;
+
+    bool lower = false, upper = false;
+    for (size_t i = 0; i < str.length; ++i)
+    {
+        ubyte c = str[i];
+        if (c >= 'a' && c <= 'z') lower = true;
+        else if (c >= 'A' && c <= 'Z') upper = true;
+        else if (c < 33 || c > 126) return null;
+    }
+    if (lower && upper)
+        return null;
+
+    auto pos = indexOf(str, '1');
+    if (str.length > 90 || pos == -1 || pos == 0 || pos + 7 > str.length)
+        return null;
+
+    ubyte[] values;
+    values.length = str.length - 1 - pos;
+    for (size_t i = 0; i < str.length - 1 - pos; ++i)
+    {
+        ubyte c = str[i + pos + 1];
+        byte rev = CHARSET_REV[c];
+
+        if (rev == -1)
+            return null;
+
+        values[i] = rev;
+    }
+    string hrp;
+    for (size_t i = 0; i < pos; ++i)
+        hrp ~= toLower(str[i]);
+
+    if (!verify_checksum(hrp, values))
+        return null;
+    if (values == null)
+        return null;
+
+    ubyte[] conv;
+    if (!convertBits(conv, values[0 .. $ - 6], 5, 8, false))
+        return null;
+
+    return conv;
+}
+
+unittest
+{
+    ubyte[] expected = [3, 3, 3, 0, 2, 15, 1];
+    auto expanded = expand_hrp("boa");
+    assert(expanded == expected);
+}
+
+unittest
+{
+    import std.stdio;
+
+    static struct ValidAddressData
+    {
+        public string address;
+        public Point pubkey;
+    }
+
+    ValidAddressData[] valid_addresses = [
+        // CoinNet Genesis Address
+        // GCOQEOHAUFYUAC6G22FJ3GZRNLGVCCLESEJ2AXBIJ5BJNUVTAERPLRIJ
+        ValidAddressData(
+            "boa1xzwsyw8q59c5qz7x669fmxe3dtx4zztyjyf6qhpgfapfd54nqy302ft5v67",
+            Point(cast(ubyte[])[
+                0x9D, 0x02, 0x38, 0xE0, 0xA1, 0x71, 0x40, 0x0B,
+                0xC6, 0xD6, 0x8A, 0x9D, 0x9B, 0x31, 0x6A, 0xCD,
+                0x51, 0x09, 0x64, 0x91, 0x13, 0xA0, 0x5C, 0x28,
+                0x4F, 0x42, 0x96, 0xD2, 0xB3, 0x01, 0x22, 0xF5])
+        ),
```

Comment:
Those are self-referential. Would be better to use the reference implementation tests here: https://github.com/sipa/bech32/blob/master/ref/c%2B%2B/tests.cpp

Reply:
Can you explain more about what you mean?

Reply:
Those are not good tests, because they are based on the implementation, rather than written independently.
Writing an implementation then adding tests that match it doesn't confirm that we conform to the standard.
Instead, the D implementation should be checked against the same tests that the C++ implementation is using.

Reply:
Added more test cases.